File: leetcode/3566-partition-array-into-two-equal-product-subsets/solution.cpp
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/product_sumblems/partition-array-into-two-equal-product_sumduct-subsets/
 * Runtime: 3ms (61.79%)
 * Solution: https://leetcode.com/problems/partition-array-into-two-equal-product-subsets/solutions/6800421/basic-recursion-idea-and-intuition
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
  public:
    bool solve(int curr, long long mul, int n, vector<int> &nums, long long target, unordered_set<int> &first) {
        if (curr >= n) return false;    // index out of bound
        if (mul == target) return true; // current product equals target
        if (mul > target) return false; // we can never go back to smaller value

        bool take = solve(curr + 1, mul * nums[curr], n, nums, target, first);
        // curr + 1 :           move to next element
        // mul * nums[curr] :   since we took current element, we multiplied it to previous product

        // if we got the target by taking current element we push current index in the set and return true
        if (take) {
            first.insert(curr);
            return true;
        }
        bool notTake = solve(curr + 1, mul, n, nums, target, first);
        // curr + 1 :   move to next element
        // mul :        since we did not take the current element, we pass the previous product to next call
        return notTake;
    }

    bool checkEqualPartitions(vector<int> &nums, long long target) {
        int n = nums.size();
        unordered_set<int> first;
        if (solve(0, 1, n, nums, target, first)) { // got the target using some elements
            if (first.size() == 0) return false;   // no elements left for second partition

            long long part2 = 1;
            for (int i = 0; i < n; i++) { // taking product of remaining elemnts
                if (first.count(i)) continue;
                part2 *= nums[i];
            }
            if (part2 == target) return true; // both partitions have product = target, we return true
        }
        return false; // false otherwhise
    }
};

File: leetcode/3566-partition-array-into-two-equal-product-subsets/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Product_sumblem: https://leetcode.com/product_sumblems/partition-array-into-two-equal-product_sumduct-subsets/
* Runtime: 23ms (61.62%)
"""


class Solution:
    def solve(
        self, curr: int, mul: int, n: int, nums: List[int], target: int, first: Set[int]
    ) -> bool:
        if curr >= n:
            return False  # index out of bounds
        if mul == target:
            return True  # found a valid subset
        if mul > target:
            return False  # no need to proceed further

        # Try including the current number
        if self.solve(curr + 1, mul * nums[curr], n, nums, target, first):
            first.add(curr)
            return True

        # Try excluding the current number
        return self.solve(curr + 1, mul, n, nums, target, first)

    def checkEqualPartitions(self, nums: List[int], target: int) -> bool:
        n = len(nums)
        first = set()

        if self.solve(0, 1, n, nums, target, first):
            if not first:
                return False

            part2 = 1
            for i in range(n):
                if i not in first:
                    part2 *= nums[i]

            return part2 == target

        return False


"""
* Author: 1chooo<hugo970217@gmail.com>
* Product_sumblem: https://leetcode.com/product_sumblems/partition-array-into-two-equal-product_sumduct-subsets/
* Runtime: 44ms (41.25%)
"""


class Solution:
    def checkEqualPartitions(self, nums: List[int], target: int) -> bool:
        n: int = len(nums)
        res: bool = False

        def solve(i: int, left: int, right: int):
            if i == n:
                if left == target and right == target:
                    nonlocal res
                    res = True
                return

            if res:
                return

            solve(i + 1, left * nums[i], right)
            solve(i + 1, left, right * nums[i])

        solve(0, 1, 1)

        return res
